mobile/test/common/integration/client_integration_test.cc

```diff
@@ -512,7 +530,40 @@ TEST_P(ClientIntegrationTest, BasicReset) {
   ASSERT_EQ(cc_.on_headers_calls, 0);
 }
 
-TEST_P(ClientIntegrationTest, BasicCancel) {
+TEST_P(ClientIntegrationTest, CancelBeforeRequestHeadersSent) {
+  autonomous_upstream_ = false;
+  initialize();
+  ConditionalInitializer headers_callback;
+
+  stream_->cancel();
+
+  terminal_callback_.waitReady();
+
+  ASSERT_EQ(cc_.on_cancel_calls, 1);
+}
+
+TEST_P(ClientIntegrationTest, CancelAfterRequestHeadersSent) {
+  initialize();
+
+  default_request_headers_.addCopy(AutonomousStream::RESPOND_AFTER_REQUEST_HEADERS, "yes");
+
+  stream_->sendHeaders(envoyToMobileHeaders(default_request_headers_), false);
+  stream_->cancel();
+  terminal_callback_.waitReady();
+  ASSERT_EQ(cc_.on_cancel_calls, 1);
+}
+
+TEST_P(ClientIntegrationTest, CancelAfterRequestComplete) {
+  autonomous_upstream_ = false;
+  initialize();
+
+  stream_->sendHeaders(envoyToMobileHeaders(default_request_headers_), true);
+  stream_->cancel();
```

Comment:
Are we guaranteed that the request has completed by the time this stream_->cancel() is called?

Reply:
yeah - sendHeaders with end stream completes the request